miio/curtain_babai.py

```diff
@@ -0,0 +1,201 @@
+    "target_position": {"siid": 2, "piid": 3},  # Range: [0, 100, 1]
+    "mode": {"siid": 2, "piid": 4},  # 0 - Normal,  1 - Reversal, 2 - Calibrate
+}
+
+# Model: "OnViz Curtain Controller (Wi-Fi)"
+MODEL_CURTAIN_BABAI = "babai.curtain.190812"
+
+
+class MotorControl(Enum):
+    Pause = 0
+    Open = 1
+    Close = 2
+
+
+class Mode(Enum):
+    Normal = 0
+    Reverse = 1
+    Calibrate = 2
+
+
+class CurtainStatus(DeviceStatus):
+    def __init__(self, data: Dict[str, Any]) -> None:
+        """Response from device.
+
+        {'motor_control': 0}
+        """
+        self.data = data
+
+    @property
+    def status(self) -> MotorControl:
+        """Device status."""
+        return MotorControl(self.data["motor_control"])
+
+    @property
+    def motor_control(self) -> MotorControl:
+        """Device status."""
+        return MotorControl(self.data["motor_control"])
+
+    @property
+    def is_open(self) -> bool:
+        """Device status."""
+        return self.motor_control == MotorControl.Open
+
+    @property
+    def is_close(self) -> bool:
+        """Device status."""
+        return self.motor_control == MotorControl.Close
+
+    @property
+    def mode(self) -> Mode:
+        """Motor rotation polarity."""
+        return Mode(self.data["mode"])
+
+    @property
+    def current_position(self) -> int:
+        """Run time of the motor."""
+        return self.data["current_position"]
+
+    @property
+    def target_position(self) -> int:
+        """Target curtain position."""
+        return self.data["target_position"]
+
+
+class CurtainBabai(MiotDevice):
+    """Main class representing the babai.curtain.190812 curtain."""
+
+    mapping = _MAPPING
+
+    @command(
+        default_output=format_output(
+            "",
+            "Motor control: {result.motor_control}\n"
+            "Motor mode: {result.mode}\n"
+            "Current position: {result.current_position}\n"
+            "Target position: {result.target_position}\n",
+        )
+    )
+    def status(self) -> CurtainStatus:
+        """Retrieve properties."""
+        # if use get_properties_for_mapping and max_len>1 then code -4004.
+        data = []
+        for k, v in self.mapping.items():
+            if "aiid" not in v:
+                data.extend(self.get_property(k))
+                # data[k] = prop["value"] if prop["code"] == 0 else None
+        return CurtainStatus(self._extract_property_value(data))
+
+    def get_property(self, name: str) -> List[Dict[str, Any]]:
+        if name not in self.mapping:
+            raise ValueError("key %s not in MAPING", name)
+        v = self.mapping[name]
+        try:
+            prop = self.get_property_by(siid=v["siid"], piid=v["piid"])
+        except DeviceError as e:
+            if e.code != -9999:
+                raise
+            sleep(5)
+            prop = self.get_property_by(siid=v["siid"], piid=v["piid"])
+        for p in prop:
+            p["did"] = name
+        return prop
```

Comment:
There should be no manual handling of this sort, it's better to find a proper solution and use that instead. Sleeping for five seconds sounds also exaggerated.